### PR TITLE
Prevent segfault errors when the SatDiagnConc diagnostic is requested; Add SatDiagnEdge collection

### LINKED ISSUE #1805: SatDiagn problems in gcclassic  CO2 run (v14.1)
In my tests for  v14.1 CO2  simulation,  all runs  go smoothly when SatDiagn is off. However  when  I turn on SatDiagn,  various problems appears.  Here is one example setting for  HISTORY.rc :
.....
  SatDiagn.template:          '%y4%m2%d2_%h2%n2z.nc4',
  SatDiagn.format:            'CFIO',
  SatDiagn.frequency:        00000001 000000
  SatDiagn.duration:          00000001 000000
  SatDiagn.hrrange:           11.98 15.02
  SatDiagn.mode:              'time-averaged'
  SatDiagn.fields:            'SatDiagnConc_CO2      ',
.....

The simulation will crash at end of first day with  segmentation error:

 if I also included  Pressure Edge to to the collection

'SatDiagnPEdge                  ' 

, I got an error message to complain ' no OH field included in the simulation '

I am not sure whether other people have met similar problem..

### PR DESCRIPTION
**Pull request by yantosca**
### Name and Institution (Required)

Name: Bob Yantosca
Institution: Harvard+GCST

### Confirm you have reviewed the following documentation

- [x] [Contributing guidelines](https://geos-chem.readthedocs.io/en/stable/reference/CONTRIBUTING.html)

### Describe the update

This is the companion PR to issue #1805, in which @lfeng89089 reported that simulations requesting the `SatDiagnConc` diagnostic field ended with a segfault error.  The root cause of this was that the `State_Diag%Archive_SatDiagnConc` was not included in the statement where `State_Diag%Archive_SatDiagn` is determined to be true or false.  Long story short, this side effect prevented the `State_Diag%SatDiagCount` array (which contains the # of updates per grid box) from being allocated when the `SatDiagnConc` diagnostic field was requested in the `SatDiagn` diagnostic collection.

We have made the following updates.

1. Make sure that the statement where `State_Diag%SatDiagn` is set to true or false uses all currently-defined `State_Diag%Archive_SatDiagn*` fields.

2. Fixed a bug where an error would be thrown in `GeosCore/diagnostics_mod.F90` if OH was not a species.  Now we restrict this error trap to the carbon and fullchem simulations (which use OH).

3. Added the `SatDiagnEdge` collection to all GEOS-Chem Classic `HISTORY.rc` template files.  This allows the user to archive e.g. `SatDiagnPEDGE`, which is placed on level edges.  Level edge fields must be saved to as separate collection than fields placed on level centesr, as there can only be one level dimension per netCDF file.

### Expected changes

This will be a zero-diff update as it only affects optional diagnostics.

### Reference(s)

N/A

### Related Github Issue(s)

Closes #1805.  This can go into a patch version such as 14.2.1.

### CODE REVIEW

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #7a6ef58 GEOS-Chem submodule update: Fix GCHP time_mod start and elapsed times
GEOS-Chem #8295fcf08 Add SatDiagnEdge collection to all GCClassic HISTORY.rc templates
HEMCO     #98adbe2 Update CHANGELOG.md

Using 24 OpenMP threads
Number of execution tests: 26

Submitted as SLURM job: 55396782
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_05x0625_NA_47L_merra2_CH4........................Execute Simulation....PASS
gc_05x0625_NA_47L_merra2_fullchem...................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem..........................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem_TOMAS15..................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem_TOMAS40..................Execute Simulation....PASS
gc_4x5_merra2_aerosol...............................Execute Simulation....PASS
gc_4x5_merra2_carbon................................Execute Simulation....PASS
gc_4x5_merra2_CH4...................................Execute Simulation....PASS
gc_4x5_merra2_CO2...................................Execute Simulation....PASS
gc_4x5_merra2_fullchem..............................Execute Simulation....PASS
gc_4x5_merra2_fullchem_aciduptake...................Execute Simulation....PASS
gc_4x5_merra2_fullchem_APM..........................Execute Simulation....PASS
gc_4x5_merra2_fullchem_benchmark....................Execute Simulation....PASS
gc_4x5_merra2_fullchem_complexSOA...................Execute Simulation....PASS
gc_4x5_merra2_fullchem_complexSOA_SVPOA.............Execute Simulation....PASS
gc_4x5_merra2_fullchem_LuoWd........................Execute Simulation....PASS
gc_4x5_merra2_fullchem_marinePOA....................Execute Simulation....PASS
gc_4x5_merra2_fullchem_RRTMG........................Execute Simulation....PASS
gc_4x5_merra2_Hg....................................Execute Simulation....PASS
gc_4x5_merra2_metals................................Execute Simulation....PASS
gc_4x5_merra2_POPs_BaP..............................Execute Simulation....PASS
gc_4x5_merra2_tagCH4................................Execute Simulation....PASS
gc_4x5_merra2_tagCO.................................Execute Simulation....PASS
gc_4x5_merra2_tagO3.................................Execute Simulation....PASS
gc_4x5_merra2_TransportTracers......................Execute Simulation....PASS
gc_4x5_merra2_TransportTracers_LuoWd................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 26
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by yantosca** [comment]
All GCHP integration tests also passed:
```console
==============================================================================
GCHP: Execution Test Results

GCClassic #8747f23 Merge PR #312 (branch 'patch-2' of github.com:sdeastham/GCHP) into dev/14.2.0
GEOS-Chem #8295fcf08 Add SatDiagnEdge collection to all GCClassic HISTORY.rc templates
HEMCO     #98adbe2 Update CHANGELOG.md

Number of execution tests: 5

Submitted as SLURM job: 55396942
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_merra2_fullchem................................Execute Simulation....PASS
gchp_merra2_fullchem_benchmark......................Execute Simulation....PASS
gchp_merra2_fullchem_RRTMG..........................Execute Simulation....PASS
gchp_merra2_tagO3...................................Execute Simulation....PASS
gchp_merra2_TransportTracers........................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 5
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by lfeng89089** [comment]
Hi Bob,

 Many thanks!!
 I copy the suggested changes to the two files of GC v14.1 , and now SatDiagn works!

Cheers
Liang


________________________________

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol [hunk 1]
```diff
@@ -292,7 +293,7 @@ COLLECTIONS: 'Restart',
 #==============================================================================
 # %%%%% THE SatDiagn COLLECTION %%%%%
 #
-# GEOS-Chem data during satellite overpass
+# GEOS-Chem data (on level edges) during satellite overpass
```
Comment (msulprizio) [inline run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol:296]: Remove `(on level edges)` here

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol [hunk 2]
```diff
@@ -332,6 +333,21 @@ COLLECTIONS: 'Restart',
                               'SatDiagnLoss_?LOS?             ',
 ::
 #==============================================================================
+# %%%%% THE SatDiagnEdge COLLECTION %%%%%
+#
+# GEOS-Chem data during satellite overpass
```
Comment (msulprizio) [inline run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol:338]: Add `(on level edges)` here

**Review by msulprizio — CHANGES_REQUESTED** [review]
I think there is a typo in the comments of HISTORY.rc.aerosol (mixing up SatDiagn and SatDiagnEdge). After fixing that, this looks good to merge.

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol (hunk 1 above)
Comment (yantosca) [reply run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol]: @msulprizio: now fixed in commit d33d1b1

File: run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol (hunk 2 above)
Comment (yantosca) [reply run/GCClassic/HISTORY.rc.templates/HISTORY.rc.aerosol]: @msulprizio: now fixed in commit d33d1b1

**Review by msulprizio — APPROVED** [review]

**Comment by yantosca** [comment]
This PR has been merged up to 14.2.1 (locally).  Integration tests are in progress.

**Comment by yantosca** [comment]
After merging on top of PR #1857, all GCHP integration tests passed:
```console
==============================================================================
GCHP: Execution Test Results

GCClassic #1ba2b3d Merge PR #1857 (Update minimum KPP version for fullchem to 3.0.0)
GEOS-Chem #5487d702c Merge PR #1808 (Prevent segfaults in SatDiagn; add SatDiagnEdge)
HEMCO     #24cf7e0 PR #215 post-merge fixes: Update CHANGELOG.md & version numbers

Number of execution tests: 5

Submitted as SLURM job: 61994565
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_merra2_fullchem................................Execute Simulation....PASS
gchp_merra2_fullchem_benchmark......................Execute Simulation....PASS
gchp_merra2_fullchem_RRTMG..........................Execute Simulation....PASS
gchp_merra2_tagO3...................................Execute Simulation....PASS
gchp_merra2_TransportTracers........................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 5
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
Also, all integration tests show zero-differences w/r/t PR #1857:
```console
Checking gchp_merra2_fullchem
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gchp_merra2_fullchem_benchmark
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gchp_merra2_fullchem_RRTMG
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gchp_merra2_tagO3
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gchp_merra2_TransportTracers
   -> No differences in OutputDir
   -> No differences in Restarts
```

**Comment by yantosca** [comment]
After merging on top of PR #1857, all GEOS-Chem Classic integration tests passed (except for TOMAS, which is a known issue that will be fixed later).
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #93e58a2 Merge PR #1857 (Update minimum KPP version for fullchem to 3.0.0)
GEOS-Chem #5487d702c Merge PR #1808 (Prevent segfaults in SatDiagn; add SatDiagnEdge)
HEMCO     #faaddbb Merge PR #220 (branch 'feature/add_geosit_as_met_option') into dev/3.7.0

Using 24 OpenMP threads
Number of execution tests: 26

Submitted as SLURM job: 61994164
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_05x0625_NA_47L_merra2_CH4........................Execute Simulation....PASS
gc_05x0625_NA_47L_merra2_fullchem...................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem..........................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem_TOMAS15..................Execute Simulation....FAIL
gc_4x5_47L_merra2_fullchem_TOMAS40..................Execute Simulation....FAIL
gc_4x5_merra2_aerosol...............................Execute Simulation....PASS
gc_4x5_merra2_carbon................................Execute Simulation....PASS
gc_4x5_merra2_CH4...................................Execute Simulation....PASS
gc_4x5_merra2_CO2...................................Execute Simulation....PASS
gc_4x5_merra2_fullchem..............................Execute Simulation....PASS
gc_4x5_merra2_fullchem_aciduptake...................Execute Simulation....PASS
gc_4x5_merra2_fullchem_APM..........................Execute Simulation....PASS
gc_4x5_merra2_fullchem_benchmark....................Execute Simulation....PASS
gc_4x5_merra2_fullchem_complexSOA...................Execute Simulation....PASS
gc_4x5_merra2_fullchem_complexSOA_SVPOA.............Execute Simulation....PASS
gc_4x5_merra2_fullchem_LuoWd........................Execute Simulation....PASS
gc_4x5_merra2_fullchem_marinePOA....................Execute Simulation....PASS
gc_4x5_merra2_fullchem_RRTMG........................Execute Simulation....PASS
gc_4x5_merra2_Hg....................................Execute Simulation....PASS
gc_4x5_merra2_metals................................Execute Simulation....PASS
gc_4x5_merra2_POPs_BaP..............................Execute Simulation....PASS
gc_4x5_merra2_tagCH4................................Execute Simulation....PASS
gc_4x5_merra2_tagCO.................................Execute Simulation....PASS
gc_4x5_merra2_tagO3.................................Execute Simulation....PASS
gc_4x5_merra2_TransportTracers......................Execute Simulation....PASS
gc_4x5_merra2_TransportTracers_LuoWd................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 24
Execution tests failed: 2
Execution tests not yet completed: 0
```
Also, most difference tests are identical w/r/t PR #1857.  The exceptions are the simulations with parallelization issues noted in PR #1682 (Hg, APM, fullchem_LuoWD). The TOMAS tests failed in both this PR and #1857, which is why it shows no differences.
```console
Checking gc_05x0625_NA_47L_merra2_CH4
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_05x0625_NA_47L_merra2_fullchem
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_47L_merra2_fullchem
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_47L_merra2_fullchem_TOMAS15
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_47L_merra2_fullchem_TOMAS40
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_aerosol
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_carbon
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_CH4
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_CO2
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_aciduptake
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_APM
   -> 3 differences found in OutputDir
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.Metrics.20190701_0000z.nc4 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.Metrics.20190701_0000z.nc4 
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.SpeciesConc.20190701_0000z.nc4 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/GEOSChem.SpeciesConc.20190701_0000z.nc4 
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/HEMCO_diagnostics.201907010000.nc 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_fullchem_APM/OutputDir/HEMCO_diagnostics.201907010000.nc 
   -> 1 difference found in Restarts
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_fullchem_APM/Restarts/GEOSChem.Restart.20190701_0100z.nc4 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_fullchem_APM/Restarts/GEOSChem.Restart.20190701_0100z.nc4 

Checking gc_4x5_merra2_fullchem_benchmark
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_complexSOA
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_complexSOA_SVPOA
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_LuoWd
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_marinePOA
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_fullchem_RRTMG
   -> 1 difference found in OutputDir
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_fullchem_RRTMG/OutputDir/GEOSChem.RRTMG.20190701_0000z.nc4 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_fullchem_RRTMG/OutputDir/GEOSChem.RRTMG.20190701_0000z.nc4 
   -> No differences in Restarts

Checking gc_4x5_merra2_Hg
   -> 1 difference found in OutputDir
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_Hg/OutputDir/GEOSChem.SpeciesConc.20190101_0000z.nc4 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_Hg/OutputDir/GEOSChem.SpeciesConc.20190101_0000z.nc4 
   -> 1 difference found in Restarts
      * GCC_14.2.1_r6/rundirs/gc_4x5_merra2_Hg/Restarts/GEOSChem.Restart.20190101_0100z.nc4 
        GCC_14.2.1_r7/rundirs/gc_4x5_merra2_Hg/Restarts/GEOSChem.Restart.20190101_0100z.nc4 

Checking gc_4x5_merra2_metals
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_POPs_BaP
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_tagCH4
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_tagCO
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_tagO3
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_TransportTracers
   -> No differences in OutputDir
   -> No differences in Restarts

Checking gc_4x5_merra2_TransportTracers_LuoWd
   -> No differences in OutputDir
   -> No differences in Restarts
```
